Fix Wither builder type resolution for nested records fix #366

araysse opened the pull request:
Root Cause

When using @Wither and @Builder annotations together on nested records (records defined inside a class), the generated wither interface incorrectly references the builder class name, causing compilation errors.

Approach

- Derive the builder’s package from the binary name (the string with $ for inner classes), not the canonical name.

- Compute the package as the substring of recordType.getName() up to the last '.'.

- Build the builder simple name using the binary part if the record is inner:

  - If inner: FooBarBuilder
  - Else: FooBuilder

- Compose the final FQN as:
  - builderClassName = pkg.isEmpty() ? builderSimpleName : pkg + "." + builderSimpleName

- Preserve generics by using TypeDef.parameterized when the record type is parameterized.

Verification

- Reproduced failure with a nested record using both annotations; after applying the change, build succeeds.

Closes: #366

araysse committed:
Fix Wither builder type resolution for nested records

araysse committed:
add jvm target 17

araysse committed:
add jvm target 17

dstepanov commented:
Please add a test that is reproducing this problem

dstepanov commented:
Why did you add JVM_17?

sdelamo reviewed (changes requested):
The pull request misses tests. Is there an issue this PR aims to fix?

sdelamo commented on sourcegen-generator-kotlin/build.gradle.kts:
```diff
@@ -16,3 +16,9 @@ dependencies {
+    compilerOptions {
+        jvmTarget = org.jetbrains.kotlin.gradle.dsl.JvmTarget.JVM_17
+    }
+}
```
```suggestion
```

sdelamo commented on test-suite-kotlin/build.gradle.kts:
```diff
@@ -25,3 +25,8 @@ dependencies {
+    compilerOptions {
+        jvmTarget = org.jetbrains.kotlin.gradle.dsl.JvmTarget.JVM_17
+    }
+}
```
```suggestion
```

araysse committed:
Update sourcegen-generator-kotlin/build.gradle.kts

Co-authored-by: Sergio del Amo <sergio.delamo@softamo.com>

araysse committed:
Update test-suite-kotlin/build.gradle.kts

Co-authored-by: Sergio del Amo <sergio.delamo@softamo.com>

araysse commented:
> Why did you add JVM_17?

i just need to make the build pass because i was have
Inconsistent JVM-target compatibility detected for tasks 'compileJava' (17) and 'compileKotlin' (21).
this pr fix the issue [#366](https://github.com/micronaut-projects/micronaut-sourcegen/issues/366)

araysse committed:
adding requested tests

# Conflicts:
#	sourcegen-generator-kotlin/build.gradle.kts
#	test-suite-kotlin/build.gradle.kts

sdelamo commented on sourcegen-generator/src/test/groovy/io/micronaut/sourcegen/generator/visitors/WitherAnnotationVisitorSpec.groovy:
```diff
@@ -0,0 +1,67 @@
+        public class Foo {
+            @Wither
+            @Builder
+            protected record Bar(String a, String b, String c) implements Foo\$BarWither {
```
users should not write code coupled to internal classes such as `Foo\$BarWither`

dstepanov replied:
How do you add Wither then? Why does it add `$` for the inner classes?

araysse replied:
Add Wither by annotating the record with @Wither (and @Builder if you want a builder). The annotation processor generates the companion Wither interface; user code should not implement it manually.

dstepanov replied:
Did you answer my question with AI?
Withers should be used by adding it as implements interface like it was before so the previous example was correct. It's just I don't think we should generate `$`, it should be `implements Foo.BarWither`

araysse committed:
fix requested test

araysse committed:
fix requested changes: no $ in generated types

araysse committed:
add jvm_17
